proposals/p1839.md

```diff
@@ -0,0 +1,331 @@
+# Multidimensional arrays
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/1839)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Problem
+
+Multidimensional arrays are actively used in numerical methods, machine
+intelligence and data science. This is one feature than makes modern Fortran
+more attractive than C++ when it comes to a choice of compiled language:
+currently, C++ lacks support of multidimensional arrays. Having Carbon implement
+this would give it a major boost in the eyes of the scientific community.
+
+Nested arrays may look as a good alternative of multidimensional arrays but
+their performance may be not effective due to splitting in memory.
+
+## Background
+
+Multidimensional array is an array with more than two dimensions which is
+continuous in memory.
+
+Multidimensional array may be stored in memory in
+[row- or column- major order](https://en.wikipedia.org/wiki/Row-_and_column-major_order).
+
+## Proposal
+
+We should add support of multidimensional arrays in Carbon via syntax extension
+for making code clean and simplier for reading and writing.
+
+```carbon
+  var A: [f64; 3; 4];
+  var values: i64 = 0;
+  for(Ai: auto in A[:,...]) {
+    for(Aij: auto in Ai[:,...]) {
+      Aij = values++;
+  } }
+```
+or
+```carbon
+  var A: [f64; 3; 4];
+  var values: i64 = 0;
+  for(i: auto in (0:2)) {
+    for(j: auto in (0:3)) {
+      A[i,j] = values++;
+  } }
+```
+
+## Details
+
+### Definition
+
+Define an array 3x2 with explicit dimensions:
+```carbon
+  var x: [i32; 3; 2] = ((0, 1, 2), (3, 4, 5));
+```
+For avoiding explicit difinition of array dimensions, use `:`:
+```carbon
+  var x: [i32; :; :] = ((0, 1, 2), (3, 4, 5));
+```
+
+#### Automatic allocation
+
+Arrays can be automatically allocated:
+```carbon
+  var x: [i32; :, :];
+```
+For avoiding Undefined Behavior, `x` has shape `(0, 0)`.
```

Comment:
It means that dimensions may vary at run-time. The most simple example is string (C# example):
```csharp
string s = "Hello";
s += ", world!";
```
In Fortran, it oftenly used for allocating arrays when sizes are becoming known:
```fortran
type(atom) :: atoms(:)
integer :: Natoms
Natoms = get_atoms_len()
allocate(atoms(Natoms))
```
In carbon, for simplifying (Sorry for Fortran-style):
```carbon
var n_atoms: i64;
var atoms: [atom; :];
n_atoms = get_atoms_len();
allocate(atoms, shape = ( n_atoms ))
```
It can be rewritten as:
```carbon
var n_atoms: i64 = get_atoms_len();
var atoms: [atom; n_atoms];
```
But I would prefer to have such constructions since they may be actively used for class fields.

Reply:
That seems like it would make the type of `var x: [i32; 3, 2] = ((0, 1, 2), (3, 4, 5));` very different from the type of `var y: [i32; :, :] = ((0, 1, 2), (3, 4, 5));`.
- `shape(x)` would presumably be a compile-time constant, but `shape(y)` could not be.
- `x` could store all its elements directly in the object (like a `std::array` does), but `y` has to be implemented with a pointer to heap-allocated memory.
- The generated code for `x[2, 1]` can consist of a single memory access, but I believe the generated code for `y[2][1]` requires at least two memory accesses, even if you don't count the indirection through the pointer mentioned above.
- `x` provides pointer stability, whereas `y` does not: `&x[0]` will never change, but `&y[0]` could have different values over the course of `y`'s lifetime.

Those differences are so basic that I think it would be misleading to use such similar syntax to represent them both.

Reply:
In my point of view, `var x: [i32; 3, 2] = ((0, 1, 2), (3, 4, 5));` and `var y: [i32; :, :] = ((0, 1, 2), (3, 4, 5));` are the same. In `y`, `:` is used for skipping direct mention of dimensions' lengths. So, 
- `shape(x)` and `shape(y)` would presumably be a compile-time constant.
- `x` and `y` could store all its elements directly in the object (like a `std::array` does)
- `x` and `y` provides pointer stability

I did not get why accessing for `y` is as `y[2][1]`, not `y[2,1]`. It is not a nested array, so accessing will be via single memory access. For nested array, you are right about two memory accesses.

Let me separate `var y: [i32; :, :] = ((0, 1, 2), (3, 4, 5));` and `var z: [i32; :, :];`. In the case of `y`, `:` is a syntax sugar for skipping direct mention of dimensions' lengths. For `z`, `:` means that this array can be allocated/reallocated. And for `z` array, it is true what you wrote about `y`, except (3).

Reply:
We should probably move this discussion to your 1D array proposal, but to answer your questions:

> I did not get why accessing for `y` is as `y[2][1]`, not `y[2,1]`.

Sorry, that was a typo. I've fixed it now.

> It is not a nested array, so accessing will be via single memory access. For nested array, you are right about two memory accesses.

Even if it's not nested I believe it needs two memory accesses. Let's assume the 2D array is implemented as a 1D array in row-major order. That means the element at row i and column j is located at index i + N * j in the underlying array, where N is the number of columns. So in order to access that element, I need to know the number of columns. If the number of columns can vary at run-time, that means I need to load the number of columns from memory before I can access the element. On the other hand, if the number of columns is fixed at compile time, I can avoid doing that load.